README.md
# Hyperdrive

# Pre-Release Checklist

- [ ] Rework Showcase-iOS to use Hyperdrive instead of Reactant
- [ ] Rework Showcase-tvOS to use Hyperdrive instead of Reactant
- [ ] Remove Reactant documentation in `/docs`
- [ ] Write Hyperdrive documentation in `/docs`
- [ ] Update `README.md` (this file)



---

WIP readme:

---

[![CI Status](https://img.shields.io/travis/Brightify/Hyperdrive?style=flat)](https://travis-ci.org/Brightify/Hyperdrive)
[![Version](https://img.shields.io/cocoapods/v/HyperdrivePlatform.svg?style=flat)][hyperdrive-cocoapods]
[![License](https://img.shields.io/cocoapods/l/HyperdrivePlatform.svg?style=flat)][hyperdrive-cocoapods]
[![Platform](https://img.shields.io/cocoapods/p/HyperdrivePlatform.svg?style=flat)][hyperdrive-cocoapods]



## Requirements

- Target platforms:
  - iOS 11.0+
  - macOS 10.13+
  - tvOS 11.0+
- Xcode 10.2.1+
- Swift 5.0+



## Support

Our forums at https://forums.brightify.org are open to everyone, feel free to reach us there!

For commercial support and help with deployment visit our website https://www.brightify.org and contact Mike.



## Get Started

Head to our [quick-start guide][quick-start] to learn how Hypedrive works and how can you integrate it in your applications.



## Versioning

This library uses semantic versioning. Our goal is to keep the API as stable as possible without introducing any breaking changes, so that new features can be used even in older projects by updating the framework.

We don't officially support anything but the latest available version. An exception to this rule are paying customers who receive an extended support as well as one-off patches.



## Authors

- [Tadeas Kriz](https://github.com/TadeasKriz)
- [Filip Dolnik](https://github.com/FilipDolnik)
- [Matyas Kriz](https://github.com/MatyasKriz)



---

Outdated readme:

---



Reactant is a foundation for rapid and safe iOS development. It allows you to cut down your development costs by improving reusability, testability and safety of your code, especially your UI. Check our our [quick-start guide][quick-start] to learn more.

[![CI Status](https://img.shields.io/travis/Brightify/Reactant.svg?style=flat)](https://travis-ci.org/Brightify/Reactant)
[![Version](https://img.shields.io/cocoapods/v/Reactant.svg?style=flat)][reactant-cocoapods]
[![License](https://img.shields.io/cocoapods/l/Reactant.svg?style=flat)][reactant-cocoapods]
[![Platform](https://img.shields.io/cocoapods/p/Reactant.svg?style=flat)][reactant-cocoapods]
[![Apps](https://img.shields.io/cocoapods/at/Reactant.svg?style=flat)][reactant-cocoapods]
[![Slack Status](https://swiftkit.brightify.org/badge.svg)][slack]

## Requirements

* iOS 9.0+
* Xcode 8.0+
* Swift 3.0+

**NOTE**: We have experimental support for *macOS* platform in `macOS` branch, but it's incomplete and definitely not recommended for production. Our plan is to add full `macOS` support in near future.

## Communication
Feel free to reach us on Slack! [https://swiftkit.brightify.org/][slack]

## Get Started
Head to our [quick-start guide][quick-start] to learn how Reactant works and what it can do to decrease your development costs!

## Versioning
This library uses semantic versioning. We won't introduce any breaking changes without releasing a new major version. Our main goal is to keep the API as stable as possible. We build our applications on top of Reactant as well and we absolutely hate any breaking changes.

## Authors
* Tadeas Kriz, [tadeas@brightify.org](mailto:tadeas@brightify.org)
* Matous Hybl, [matous@brightify.org](mailto:matous@brightify.org)
* Filip Dolník, [filip@brightify.org](mailto:filip@brightify.org)

## Used libraries

### Runtime

* [Result](https://github.com/antitypical/Result)
* [SnapKit](https://github.com/SnapKit/SnapKit)
* [RxSwift](https://github.com/ReactiveX/RxSwift)
* [RxCocoa](https://github.com/ReactiveX/RxSwift)
* [RxOptional](https://github.com/RxSwiftCommunity/RxOptional)
* [RxDataSources](https://github.com/RxSwiftCommunity/RxDataSources)

### Tests

* [Quick](https://github.com/Quick/Quick)
* [Nimble](https://github.com/Quick/Nimble)

[quick-start]: https://docs.hyperdrive.app/getting-started/quickstart.html
[hyperdrive-cocoapods]: https://cocoapods.org/pods/HyperdrivePlatform
[forums]: https://forums.brightify.org/

CHANGELOG.md
# Changelog

## 2.0.0
* Rename to Hyperdrive
* Remove RxSwift dependency
* Add support for layout margins

## 1.2.0
* Add takeUntil to Wireframe
* Add option to disable automatic cell deselection
* Add ActivityIndicator to subspec groups
* Add reference for create<U,T>() and rewrite()
* Deprecate ButtonBase with ControlBase
* Add system font support
* Update project for Swift 4.1
* Add generic PickerView
* Add method for stacking UIViews inside UIView
* Fix `rotate` in CGAffineTransformation extensions
* Add `dequeueAndConfigure(identifier:indexPath:factory:model:mapAction:)` to `CollectionView`
* Add better initial configuration of `TableView` and `CollectionView`
* Add styling methods for `UINavigationController` and `UITabBarController`
* Remove RxSwift and SnapKit dependencies from `Configuration` subspec
* Add Observable navigation

## 1.1.0
* Add FallbackSafeAreaInsets subspec with a basic implementation of a `safeAreaInsets` and `safeAreaLayoutGuide` fallback for iOS 10.
* Add support for tvOS
* Add reference guide
* Add more docs
* Add tutorials
* Add more tests

## 1.0.6
* Fix TableViewBase and CollectionViewBase memory leak.

## 1.0.5
* Add create to Wireframe with controller result helper
* Add default implementation to DialogDismissalController
* Add option to have present dialog with result in UINavigationController
* Add styling for DialogControllerBase's `view`
* Possibly breaking: changed `bind(items: [MODEL])` to `bind(items: Observable<[MODEL]>)` in both `TableViewBase` and `CollectionViewBase`. This change was made because RxSwift changed the internals of delegates and dataSources and each `update` caused the TableView/CollectionView to scroll to the beginning.

## 1.0.4
* Improved documentation

## 1.0.3
* Fixed `TextField` where placeholder didn't have correct position

## 1.0.2
* Added `setBackgroundColor:forState` objc alias for ReactantUI

## 1.0.1
* Fixed Reactant Example project

## 0.6.0

* Complete API redesign
* Added tests
* Added documentation
* Added changelog


Platform/Project.swift
import ProjectDescription
import Foundation

enum HyperdrivePlatform: CaseIterable {
    case iOS
    case macOS
    case tvOS

    var name: String {
        switch self {
        case .iOS:
            return "iOS"
        case .macOS:
            return "macOS"
        case .tvOS:
            return "tvOS"
        }
    }

    var platform: Platform {
        switch self {
        case .iOS:
            return .iOS
        case .tvOS:
            return .tvOS
        case .macOS:
            return .macOS
        }
    }

    var deploymentTarget: DeploymentTarget? {
        switch self {
        case .iOS:
            return DeploymentTarget.iOS(targetVersion: "11.0", devices: [.iphone, .ipad, .mac])
        case .macOS:
            return DeploymentTarget.macOS(targetVersion: "10.13")
        case .tvOS:
            return nil
        }
    }
}

let targets = HyperdrivePlatform.allCases.map {
    Target(
        name: "Hyperdrive-\($0.name)",
        platform: $0.platform,
        product: .framework,
        productName: "Hyperdrive",
        bundleId: "org.brightify.hyperdrive.Platform",
        deploymentTarget: $0.deploymentTarget,
        infoPlist: "Metadata/Info.plist",
        sources: [
            "Sources/**",
        ],
        resources: [
        ],
        dependencies: [
            TargetDependency.cocoapods(path: ".."),
        ],
        settings: Settings(base: [
            "OTHER_SWIFT_FLAGS": "$(inherited) -DEnableExperimentalFeatures",
        ], defaultSettings: .recommended))
}

let rxTargets = HyperdrivePlatform.allCases.map {
    Target(
        name: "RxHyperdrive-\($0.name)",
        platform: $0.platform,
        product: .framework,
        productName: "Hyperdrive",
        bundleId: "org.brightify.hyperdrive.Platform",
        deploymentTarget: $0.deploymentTarget,
        infoPlist: "Metadata/Info.plist",
        sources: [
            "Sources/**",
            "RxSources/**",
        ],
        resources: [
        ],
        dependencies: [
            TargetDependency.cocoapods(path: ".."),
        ]
    )
}

let testTargets = HyperdrivePlatform.allCases.map {
    Target(
        name: "HyperdriveTests-\($0.name)",
        platform: $0.platform,
        product: .unitTests,
        productName: "HyperdriveTests",
        bundleId: "org.brightify.hyperdrive.PlatformTests",
        deploymentTarget: $0.deploymentTarget,
        infoPlist: "Metadata/Info.plist",
        sources: [
           "Tests/CollectionView/**",
           "Tests/Configuration/**",
           "Tests/Core/**",
           "Tests/StaticMap/**",
           "Tests/StaticMap/**",
           "Tests/TableView/**",
           "Tests/TestUtils/**",
           "Tests/Validation/**",
           "Tests/Wireframe/**",
        ],
        resources: [
        ],
        dependencies: [
            TargetDependency.cocoapods(path: ".."),
        ])
}

let rxTestTargets = HyperdrivePlatform.allCases.map {
    Target(
        name: "RxHyperdriveTests-\($0.name)",
        platform: $0.platform,
        product: .unitTests,
        productName: "HyperdriveTests",
        bundleId: "org.brightify.hyperdrive.PlatformTests",
        deploymentTarget: $0.deploymentTarget,
        infoPlist: "Metadata/Info.plist",
        sources: [
           "Tests/**"
        ],
        resources: [
        ])
}

let schemes = HyperdrivePlatform.allCases.map {
    Scheme(
        name: "Hyperdrive-\($0.name)",
        shared: true,
        buildAction: BuildAction(targets: ["Hyperdrive-\($0.name)"]),
        testAction: TestAction(targets: ["HyperdriveTests-\($0.name)"]),
        runAction: nil)
}

let project = Project(
    name: "Platform",
    targets: targets + rxTargets + testTargets + rxTestTargets,
    schemes: schemes,
    additionalFiles: [
        "../HyperdrivePlatform.podspec"
    ])
